### PR TITLE
small fix to optimization code

### PR DESCRIPTION
**Pull request by mavenlin**
1. add nesterov to SGD
2. add weight decay to gamma of batchnorm

### CODE REVIEW

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -251,9 +251,10 @@ class SGD(Optimizer):
     param_idx2name : dict of string/int to float, optional
         special treat weight decay in parameter ends with bias, gamma, and beta
     """
-    def __init__(self, momentum=0.0, **kwargs):
+    def __init__(self, momentum=0.0, nesterov=False, **kwargs):
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: split to a new one?

File: python/mxnet/optimizer.py [hunk 2]
```diff
@@ -157,7 +157,7 @@ def set_wd_mult(self, args_wd_mult):
         """
         self.wd_mult = {}
         for n in self.idx2name.values():
-            if not n.endswith('_weight'):
+            if not (n.endswith('_weight') or n.endswith('_gamma')):
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: Why should gamma be decayed?

File: python/mxnet/optimizer.py (hunk 1 above)
Comment (mavenlin) [inline python/mxnet/optimizer.py]: I think it is compatible with the original sgd, can you double check that when nesterov is false it is same with the original code?

File: python/mxnet/optimizer.py (hunk 2 above)
Comment (mavenlin) [inline python/mxnet/optimizer.py]: I think decay for gamma is a necessary regularization, you can get better generalization.

Comment (piiswrong) [inline python/mxnet/optimizer.py]: gamma is variance right? Still feels wired that it should be decayed.

Comment (vchuravy) [inline python/mxnet/optimizer.py]: @mavenlin Do you have a reference for decaying gamma? Or an example of how other libraries do this?

**Comment by tqchen** [comment]
any update on this? we can consider merge it in if the optimizer is backward compatible with original SGD when nestrov is false

File: python/mxnet/optimizer.py (hunk 2 above)
Comment (superzrx) [inline python/mxnet/optimizer.py]: @piiswrong @mavenlin @vchuravy 
decaying is applied on weight and bias( gamma and beta is regarded as weight and bias ) in torch
reference
https://github.com/torch/optim/blob/master/sgd.lua
https://github.com/torch/nn/blob/master/BatchNormalization.lua
https://github.com/facebook/fb.resnet.torch/

**Comment by tqchen** [comment]
I am going to merge this in 24 hours. Please let me know if there is any concerns

File: python/mxnet/optimizer.py (hunk 2 above)
Comment (piiswrong) [inline python/mxnet/optimizer.py]: torch aggregate all params togather. Decaying bias is definitely not normal

File: python/mxnet/optimizer.py [hunk 3]
```diff
@@ -299,8 +300,13 @@ def update(self, index, weight, grad, state):
         if state:
             mom = state
             mom[:] *= self.momentum
-            mom[:] += -lr * (grad + wd * weight)
-            weight[:] += mom
+            grad += wd * weight
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: modifying gradient array can cause problems. It's better to use a tmp variable

File: python/mxnet/optimizer.py (hunk 1 above)
Comment (piiswrong) [inline python/mxnet/optimizer.py]: I think this should be split into a new optimizer. Merging makes the code more complicated and harder to understand. It also makes it harder for users to find optimizers to use

File: python/mxnet/optimizer.py (hunk 3 above)
Comment (tqchen) [inline python/mxnet/optimizer.py]: @mavenlin can you modify this back? I think I can merge this after it is fixed

File: python/mxnet/optimizer.py (hunk 2 above)
Comment (mavenlin) [inline python/mxnet/optimizer.py]: I experimented with this, and it is better to decay gamma.

File: python/mxnet/optimizer.py (hunk 3 above)
Comment (mavenlin) [inline python/mxnet/optimizer.py]: What problem can this cause? In the previous line we've already done this:
`grad = grad * self.rescale_grad`

Comment (mavenlin) [inline python/mxnet/optimizer.py]: What problem can it cause?
Even though it could cause problem writing to the original gradient, 
in this previous line we already done this
`grad = grad * self.rescale_grad`, 
So the grad here is no longer the original grad passed to this function.

Comment (piiswrong) [inline python/mxnet/optimizer.py]: Ok, good point. Please split this to a new optimizer and we can merge

Comment (tqchen) [inline python/mxnet/optimizer.py]: @mavenlin Let us do this. Create a new NAGOptimizer to subclass SGDOptimizer and add your stuffs there

Comment (mavenlin) [inline python/mxnet/optimizer.py]: @tqchen what does NAG mean?

Comment (mavenlin) [inline python/mxnet/optimizer.py]: @piiswrong @tqchen There is only one line difference in nesterov vs sgd, subclassing duplicates the update code.

Comment (piiswrong) [inline python/mxnet/optimizer.py]: yes, but the problem is mixing them together makes it hard to figure out which part is which and hard to modify either of them. Also it makes it hard for new users to find nestrov because it's hidden in sgd

Comment (tqchen) [inline python/mxnet/optimizer.py]: I agree with @piiswrong that a subclassing would be better choice. and the update rule was not to long to write

**Comment by tqchen** [comment]
@mavenlin can you act upon the comments and change it to subclassing? thanks!

**Comment by mavenlin** [comment]
@tqchen @piiswrong it is fixed now

File: python/mxnet/optimizer.py (hunk 2 above)
Comment (vchuravy) [inline python/mxnet/optimizer.py]: @pluskid Is this something we should mirror in the Julia code?

Comment (pluskid) [inline python/mxnet/optimizer.py]: @vchuravy yes, this allows to set learning rate easily for each parameter. Do you have time to sync this feature to Julia? I might have some time to do some porting to julia after the end of this semester, but currently quite busy.

**Comment by tqchen** [comment]
Thanks! this is now merged
